### ISSUE TITLE
Bug in the DeadStoreEliminator

### ISSUE THREAD

**Issue by pgarba** [lifting-bits/remill#350]
Hi,

I ran into a bug with the DSE.

Input IR:

```llvm
; Function Attrs: noinline nounwind
define dso_local %struct.Memory* @FS_149464250(%struct.State* noalias dereferenceable(3376), i64, %struct.Memory* noalias returned) local_unnamed_addr #0 {
entry:
  %PC = getelementptr inbounds %struct.State, %struct.State* %0, i64 0, i32 6, i32 33, i32 0, i32 0
  %RSP = getelementptr inbounds %struct.State, %struct.State* %0, i64 0, i32 6, i32 13, i32 0, i32 0
  %3 = getelementptr inbounds %struct.State, %struct.State* %0, i64 0, i32 6, i32 11, i32 0
  %EDI.i = bitcast %union.anon* %3 to i32*
  %R14.i21 = getelementptr inbounds %struct.State, %struct.State* %0, i64 0, i32 6, i32 29, i32 0, i32 0
  %4 = load i64, i64* %R14.i21, align 8
  %5 = load i32, i32* %EDI.i, align 4
  %cf.i.i.i23 = getelementptr inbounds %struct.State, %struct.State* %0, i64 0, i32 2, i32 1
  %pf.i.i.i24 = getelementptr inbounds %struct.State, %struct.State* %0, i64 0, i32 2, i32 3
  %zf.i.i.i25 = getelementptr inbounds %struct.State, %struct.State* %0, i64 0, i32 2, i32 7
  %sf.i.i.i26 = getelementptr inbounds %struct.State, %struct.State* %0, i64 0, i32 2, i32 9
  %of.i.i.i27 = getelementptr inbounds %struct.State, %struct.State* %0, i64 0, i32 2, i32 13
  %af.i.i.i28 = getelementptr inbounds %struct.State, %struct.State* %0, i64 0, i32 2, i32 5
  %6 = trunc i64 %4 to i32
  %or3.i26.i50.i = or i32 %5, %6
  %RDI.i13 = getelementptr inbounds %struct.State, %struct.State* %0, i64 0, i32 6, i32 11, i32 0, i32 0
  %7 = load i64, i64* %RDI.i13, align 8
  %8 = trunc i64 %7 to i32
  %and3.i.i70.i = and i32 %or3.i26.i50.i, %8
  %shl.i132.i.i = shl i32 %or3.i26.i50.i, 24
  %conv.i128.i83.i = lshr i32 %or3.i26.i50.i, 9
  %or3.i124.i.i = or i32 %conv.i128.i83.i, %shl.i132.i.i
  %shr.i137.i.i = lshr i32 %or3.i26.i50.i, 8
  %9 = and i32 %shr.i137.i.i, 1
  %add.i48.i.i = add nsw i32 %9, %or3.i124.i.i
  %add.i.i89.i = add i32 %add.i48.i.i, %and3.i.i70.i
  %RDX.i = getelementptr inbounds %struct.State, %struct.State* %0, i64 0, i32 6, i32 7, i32 0, i32 0
  %.demorgan = and i32 %add.i.i89.i, %and3.i.i70.i
  %10 = xor i32 %.demorgan, -1
  %conv.i12.i15.i = zext i32 %10 to i64
  %RAX.i = getelementptr inbounds %struct.State, %struct.State* %0, i64 0, i32 6, i32 1, i32 0, i32 0
  %RBX.i = getelementptr inbounds %struct.State, %struct.State* %0, i64 0, i32 6, i32 3, i32 0, i32 0
  store i64 %conv.i12.i15.i, i64* %RDI.i13, align 8
  store i64 2944344100, i64* %RBX.i, align 8
  store i64 1820957027, i64* %RAX.i, align 8
  store i8 1, i8* %af.i.i.i28, align 1
  store i64 921427912, i64* %RDX.i, align 8
  store i8 1, i8* %cf.i.i.i23, align 1
  store i8 0, i8* %pf.i.i.i24, align 1
  store i8 0, i8* %zf.i.i.i25, align 1
  store i8 0, i8* %sf.i.i.i26, align 1
  store i8 0, i8* %of.i.i.i27, align 1
  store i64 4101712099, i64* %R14.i21, align 8
  store i64 23009984, i64* %RSP, align 8
  store i64 5503624381, i64* %PC, align 8
  %11 = tail call %struct.Memory* @__remill_jump(%struct.State* nonnull %0, i64 5503624381, %struct.Memory* %2) #3
  ret %struct.Memory* %2
}
```

After DSE:

```llvm
; Function Attrs: noinline nounwind
define dso_local %struct.Memory* @FS_149464250(%struct.State* noalias dereferenceable(3376), i64, %struct.Memory* noalias returned) local_unnamed_addr #0 {
entry:
  %3 = getelementptr inbounds %struct.State, %struct.State* %0, i64 0, i32 6, i32 33, i32 0, i32 0
  %4 = getelementptr inbounds %struct.State, %struct.State* %0, i64 0, i32 6, i32 13, i32 0, i32 0
  %5 = getelementptr inbounds %struct.State, %struct.State* %0, i64 0, i32 6, i32 11, i32 0
  %6 = bitcast %union.anon* %5 to i32*
  %7 = getelementptr inbounds %struct.State, %struct.State* %0, i64 0, i32 6, i32 29, i32 0, i32 0
  %8 = load i64, i64* %7, align 8
  %9 = load i64, i64* %19, align 8 // Bug in here! The load will access %19 which is not allowed the right variable would be %5
  %10 = trunc i64 %9 to i32
  %11 = getelementptr inbounds %struct.State, %struct.State* %0, i64 0, i32 2, i32 1
  %12 = getelementptr inbounds %struct.State, %struct.State* %0, i64 0, i32 2, i32 3
  %13 = getelementptr inbounds %struct.State, %struct.State* %0, i64 0, i32 2, i32 7
  %14 = getelementptr inbounds %struct.State, %struct.State* %0, i64 0, i32 2, i32 9
  %15 = getelementptr inbounds %struct.State, %struct.State* %0, i64 0, i32 2, i32 13
  %16 = getelementptr inbounds %struct.State, %struct.State* %0, i64 0, i32 2, i32 5
  %17 = trunc i64 %8 to i32
  %18 = or i32 %10, %17
  %19 = getelementptr inbounds %struct.State, %struct.State* %0, i64 0, i32 6, i32 11, i32 0, i32 0
  %20 = trunc i64 %9 to i32
  %21 = and i32 %18, %20
  %22 = shl i32 %18, 24
  %23 = lshr i32 %18, 9
  %24 = or i32 %23, %22
  %25 = lshr i32 %18, 8
  %26 = and i32 %25, 1
  %27 = add nsw i32 %26, %24
  %28 = add i32 %27, %21
  %29 = getelementptr inbounds %struct.State, %struct.State* %0, i64 0, i32 6, i32 7, i32 0, i32 0
  %30 = and i32 %28, %21
  %31 = xor i32 %30, -1
  %32 = zext i32 %31 to i64
  %33 = getelementptr inbounds %struct.State, %struct.State* %0, i64 0, i32 6, i32 1, i32 0, i32 0
  %34 = getelementptr inbounds %struct.State, %struct.State* %0, i64 0, i32 6, i32 3, i32 0, i32 0
  store i64 %32, i64* %19, align 8
  store i64 2944344100, i64* %34, align 8
  store i64 1820957027, i64* %33, align 8
  store i8 1, i8* %16, align 1
  store i64 921427912, i64* %29, align 8
  store i8 1, i8* %11, align 1
  store i8 0, i8* %12, align 1
  store i8 0, i8* %13, align 1
  store i8 0, i8* %14, align 1
  store i8 0, i8* %15, align 1
  store i64 4101712099, i64* %7, align 8
  store i64 23009984, i64* %4, align 8
  store i64 5503624381, i64* %3, align 8
  %35 = tail call %struct.Memory* @__remill_jump(%struct.State* nonnull %0, i64 5503624381, %struct.Memory* %2) #3
  ret %struct.Memory* %2
}
```

The bug is in the line:

```llvm
"%9 = load i64, i64* %19, align 8"
```

This code is wrong because accessing %19 is not allowed. The right variable should be %5.

Peter

**Comment by pgoodman** [comment]
Duplicate of #308 I think.

**Comment by pgarba** [comment]
I created a fix for it by introducing a check that verifies if the slot if dominating the load instruction. If not, it will update the slot.

[0001-DSE-Check-if-the-slot-is-dominating-the-load-instru.zip](https://github.com/trailofbits/remill/files/3120521/0001-DSE-Check-if-the-slot-is-dominating-the-load-instru.zip)

**Comment by pgoodman** [comment]
Fixed in #373
